src/Npgsql/HardCancellationBlock.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql
+{
+    readonly struct HardCancellationBlock : IDisposable
+    {
+        static readonly AsyncLocal<bool> _entered = new AsyncLocal<bool>();
+
+        public static bool Entered => _entered.Value;
+
+        readonly bool _outerBlock;
+
+        public HardCancellationBlock(bool outerBlock) =>_outerBlock = outerBlock;
+
+        public static IDisposable Enter()
+        {
+            var context = new HardCancellationBlock(!Entered);
+            if (!Entered)
+                _entered.Value = true;
+            return context;
+        }
```

Comment:
This one can be made shorter with a single access to `Entered` via rewriting this too:

```csharp
public HardCancellationBlock(bool entered) =>
    _outerBlock = s_entered.Value = entered;

public static HardCancellationBlock Enter() =>
    Entered
        ? new HardCancellationBlock()
        : new HardCancellationBlock(entered: true);
```

You even can replace `new HardCancellationBlock()` by `default` and write it in a single line of code. This way you will access the async local only once which is much better since it's not so cheap as a static variable. Under the hood all async locals stored in a dictionary.